File: android/app/src/main/java/com/smartexpense/android/data/repository/UserRepositoryImpl.java
package com.smartexpense.android.data.repository;

import com.smartexpense.android.data.remote.api.UserApi;
import com.smartexpense.android.data.remote.dto.request.UpdateProfileRequestDto;
import com.smartexpense.android.data.remote.dto.response.UserProfileResponseDto;
import com.smartexpense.android.domain.repository.ResultCallback;
import com.smartexpense.android.domain.repository.UserRepository;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UserRepositoryImpl implements UserRepository {

    private final UserApi api;

    public UserRepositoryImpl(UserApi api) {
        this.api = api;
    }

    @Override
    public void getMe(ResultCallback<UserProfileResponseDto> callback) {
        api.getMe().enqueue(new Callback<UserProfileResponseDto>() {
            @Override
            public void onResponse(Call<UserProfileResponseDto> call, Response<UserProfileResponseDto> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body());
                } else {
                    callback.onError("Lỗi tải thông tin: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<UserProfileResponseDto> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }

    @Override
    public void updateMe(UpdateProfileRequestDto request, ResultCallback<UserProfileResponseDto> callback) {
        api.updateMe(request).enqueue(new Callback<UserProfileResponseDto>() {
            @Override
            public void onResponse(Call<UserProfileResponseDto> call, Response<UserProfileResponseDto> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body());
                } else {
                    callback.onError("Lỗi cập nhật thông tin: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<UserProfileResponseDto> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }
}

File: android/app/src/main/kotlin/com/smartexpense/android/data/repository/UserRepositoryImpl.kt
package com.smartexpense.android.data.repository

import com.smartexpense.android.data.remote.api.UserApi
import com.smartexpense.android.data.remote.dto.request.UpdateProfileRequestDto
import com.smartexpense.android.data.remote.dto.response.UserProfileResponseDto
import com.smartexpense.android.domain.repository.UserRepository

class UserRepositoryImpl(private val api: UserApi) : UserRepository {
    override suspend fun getMe(): Result<UserProfileResponseDto> {
        return try {
            val response = api.getMe()
            Result.success(response)
        } catch (e: Exception) {
            Result.failure(Exception("Lỗi tải thông tin: ${e.message}", e))
        }
    }

    override suspend fun updateMe(request: UpdateProfileRequestDto): Result<UserProfileResponseDto> {
        return try {
            val response = api.updateMe(request)
            Result.success(response)
        } catch (e: Exception) {
            Result.failure(Exception("Lỗi cập nhật thông tin: ${e.message}", e))
        }
    }
}
